File: Utils/SWTUtils/src/main/java/org/tobi29/scapes/engine/swt/util/framework/Application.java
package org.tobi29.scapes.engine.swt.util.framework;

import java8.util.Optional;
import org.eclipse.swt.program.Program;
import org.eclipse.swt.widgets.Display;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.tobi29.scapes.engine.swt.util.platform.Platform;
import org.tobi29.scapes.engine.swt.util.platform.PlatformLinux;
import org.tobi29.scapes.engine.utils.Crashable;
import org.tobi29.scapes.engine.utils.SleepUtil;
import org.tobi29.scapes.engine.utils.io.filesystem.CrashReportFile;
import org.tobi29.scapes.engine.utils.io.filesystem.FilePath;
import org.tobi29.scapes.engine.utils.io.filesystem.FileUtil;
import org.tobi29.scapes.engine.utils.task.TaskExecutor;

import java.io.IOException;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public abstract class Application implements Runnable, Crashable {
    protected static final Platform PLATFORM;
    private static final Logger LOGGER =
            LoggerFactory.getLogger(Application.class);

    static {
        // TODO: Pick based on SWT implementation
        PLATFORM = new PlatformLinux();
    }

    protected final Display display;
    protected final TaskExecutor taskExecutor;

    protected Application(String name, String id, String version) {
        Display.setAppName(name);
        Display.setAppVersion(version);
        display = Display.getDefault();
        taskExecutor = new TaskExecutor(this, id);
    }

    public static Platform platform() {
        return PLATFORM;
    }

    public Display display() {
        return display;
    }

    public TaskExecutor taskExecutor() {
        return taskExecutor;
    }

    @Override
    public void run() {
        try {
            init();
            while (display.getShells().length > 0) {
                if (!display.readAndDispatch()) {
                    display.sleep();
                }
            }
            dispose();
            taskExecutor.shutdown();
        } catch (Throwable e) {
            crash(e);
        }
        display.dispose();
    }

    @SuppressWarnings("CallToSystemExit")
    @Override
    public void crash(Throwable e) {
        Optional<FilePath> report = writeCrash(e);
        if (!report.isPresent()) {
            System.exit(1);
            return;
        }
        FilePath path = report.get();
        if (!Program.launch(path.toString())) {
            display.asyncExec(() -> {
                Program.launch(path.toString());
                System.exit(1);
            });
            SleepUtil.sleepAtLeast(1000);
            System.exit(1);
        }
    }

    private Optional<FilePath> writeCrash(Throwable e) {
        try {
            FilePath path = FileUtil.createTempFile("CrashReport", ".txt");
            Map<String, String> debugValues = new ConcurrentHashMap<>();
            CrashReportFile
                    .writeCrashReport(e, path, "ScapesEngine", debugValues);
            return Optional.of(path);
        } catch (IOException e1) {
        }
        return Optional.empty();
    }

    protected abstract void init();

    protected abstract void dispose();
}

File: SWTUtilsJVM/src/main/kotlin/org/tobi29/scapes/engine/swt/util/framework/Application.kt
/*
 * Copyright 2012-2017 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.tobi29.scapes.engine.swt.util.framework

import kotlinx.coroutines.experimental.CommonPool
import kotlinx.coroutines.experimental.CoroutineDispatcher
import kotlinx.coroutines.experimental.Runnable
import org.eclipse.swt.SWT
import org.eclipse.swt.program.Program
import org.eclipse.swt.widgets.Display
import org.eclipse.swt.widgets.Shell
import org.tobi29.scapes.engine.swt.util.platform.*
import org.tobi29.scapes.engine.swt.util.widgets.Dialogs
import org.tobi29.scapes.engine.utils.Crashable
import org.tobi29.scapes.engine.utils.Version
import org.tobi29.scapes.engine.utils.io.*
import org.tobi29.scapes.engine.utils.io.filesystem.FilePath
import org.tobi29.scapes.engine.utils.io.filesystem.createTempFile
import org.tobi29.scapes.engine.utils.io.filesystem.write
import org.tobi29.scapes.engine.utils.logging.KLogging
import org.tobi29.scapes.engine.utils.sleepAtLeast
import kotlin.coroutines.experimental.CoroutineContext
import kotlin.system.exitProcess

abstract class Application(
        name: String,
        id: String,
        version: Version,
        val taskExecutor: CoroutineContext = CommonPool
) : CoroutineDispatcher(), Runnable, Crashable {
    val display: Display

    init {
        Display.setAppName(name)
        Display.setAppVersion(version.toString())
        display = Display.getDefault()
    }

    fun message(style: Int,
                title: String,
                message: String): Int {
        val shell = activeShell ?: return 0
        return Dialogs.openMessage(shell, style, title, message)
    }

    val activeShell: Shell?
        get() {
            val shell = display.activeShell
            if (shell == null) {
                val shells = display.shells
                if (shells.isEmpty()) {
                    return null
                }
                return shells[0]
            }
            return shell
        }

    fun access(runnable: () -> Unit) {
        display.syncExec(runnable)
    }

    fun accessAsync(runnable: () -> Unit) {
        display.asyncExec(runnable)
    }

    override fun run() {
        try {
            initApplication()
            while (!done()) {
                if (!display.readAndDispatch()) {
                    display.sleep()
                }
            }
            display.dispose()
            disposeApplication()
        } catch (e: Throwable) {
            crash(e)
        }
    }

    fun done(): Boolean {
        return display.isDisposed || display.shells.isEmpty()
    }

    fun initApplication() {
        init()
    }

    fun disposeApplication() {
        dispose()
    }

    override fun crash(e: Throwable): Nothing {
        logger.error(e) { "Application crashed:" }
        val path = writeCrash(e) ?: exitProcess(1)
        if (!Program.launch(path.toString())) {
            display.asyncExec {
                Program.launch(path.toString())
                exitProcess(1)
            }
            sleepAtLeast(1000)
        }
        exitProcess(1)
    }

    override fun dispatch(context: CoroutineContext,
                          block: Runnable) {
        display.asyncExec(block)
    }

    private fun writeCrash(e: Throwable): FilePath? {
        try {
            val path = createTempFile("CrashReport", ".txt")
            write(path) {
                it.writeCrashReport(e, "SWT Application",
                        crashReportSectionStacktrace(e),
                        crashReportSectionActiveThreads(),
                        crashReportSectionSystemProperties())
            }
            return path
        } catch (e1: IOException) {
        }
        return null
    }

    protected abstract fun init()

    protected abstract fun dispose()

    companion object : KLogging() {
        val platform: Platform

        init {
            when (SWT.getPlatform()) {
                "gtk" -> platform = PlatformLinux()
                "cocoa" -> platform = PlatformMacOSX()
                "win32" -> platform = PlatformWindows()
                else -> {
                    logger.warn { "Unknown SWT platform: ${SWT.getPlatform()}" }
                    platform = PlatformUnknown()
                }
            }
        }
    }
}
